src/node_quic_crypto.cc

```diff
@@ -713,56 +727,128 @@ Local<Value> GetServerName(
 }
 
 // Get the ALPN protocol identifier that was negotiated for the session
-Local<Value> GetALPNProtocol(Environment* env, SSL* ssl) {
+Local<Value> GetALPNProtocol(QuicSession* session) {
   Local<Value> alpn;
   const unsigned char* alpn_buf = nullptr;
   unsigned int alpnlen;
+  auto ctx = session->CryptoContext();
 
-  SSL_get0_alpn_selected(ssl, &alpn_buf, &alpnlen);
+  SSL_get0_alpn_selected(**ctx, &alpn_buf, &alpnlen);
   if (alpnlen == sizeof(NGTCP2_ALPN_H3) - 2 &&
       memcmp(alpn_buf, NGTCP2_ALPN_H3 + 1, sizeof(NGTCP2_ALPN_H3) - 2) == 0) {
-    alpn = env->quic_alpn_string();
+    alpn = session->env()->quic_alpn_string();
   } else {
-    alpn = OneByteString(env->isolate(), alpn_buf, alpnlen);
+    alpn = OneByteString(session->env()->isolate(), alpn_buf, alpnlen);
   }
   return alpn;
 }
 
-Local<Value> GetCipherName(Environment* env, SSL* ssl) {
+Local<Value> GetCertificate(QuicSession* session) {
+  crypto::ClearErrorOnReturn clear_error_on_return;
+  auto ctx = session->CryptoContext();
+  Local<Value> value = v8::Undefined(session->env()->isolate());
+  X509* cert = SSL_get_certificate(**ctx);
+  if (cert != nullptr)
+    value = crypto::X509ToObject(session->env(), cert);
+  return value;
+}
+
+Local<Value> GetEphemeralKey(QuicSession* session) {
+  Environment* env = session->env();
+  Local<Context> context = env->context();
+
+  Local<Object> info = Object::New(env->isolate());
+  auto ctx = session->CryptoContext();
```

Comment:
You don’t have to agree with me, but I still kind of don’t like `auto` here … I’d still always have to look up the type if I wanted to know it.